### ISSUE TITLE
OTA: If a third party app causes OTA failure - alert the user and give instructions

### ISSUE THREAD

**Issue by zalberico** [urbit/urbit#6285]
In the last we release we had the following happen:
- Citadel and Gora signaled support for 415 kelvin
- When the 415 base OTA was released users attempted to suspend other apps without 415 support and upgrade
- Citadel and Gora didn't actually work on 415
- The OTA makes an assumption that apps marked supported are expected by the user to work, therefore if those apps don't successfully upgrade, the system rolls back the OTA to a working state.
- There's then a bug where the system attempts to retry the OTA forever
- The user is then confused about why the upgrade failed and is looping

Ideas:
- While it's reasonable to think the user would want to revert the OTA if some important third party app they thought supported it is actually broken, I'm not sure the system should do that. If an app dev signals support for a new version but doesn't confirm it works on that new version that's on the app dev. I'm not sure we should roll back the base OTA because of this.
- Even if we do decide to roll back, we should then stop reattempting the OTA and loudly tell the user in some form: "Apps X and Y failed to upgrade despite claiming support for the new version. Suspend them and upgrade anyway?"

My preference is probably to upgrade base no matter what and if a third party app fails despite claiming support - the user will complain to the dev. I think this is less confusing/more expected than the system doing what it's doing currently.

**Comment by Fang-** [comment]
I'd argue against suspending apps if they fail to upgrade cleanly despite initially signaling the contrary. Without the user explicitly suspending the app, we have no way of knowing whether that's ok or not: it may be in active use for their business, event, or other "sensitive" use-case, which will take precedence over the system update.  
Suspending apps without first consulting the user gives big "Windows will restart in one minute to apply some updates, press [ok] or [restart now]" vibes.

We _do_ clearly need to handle such states better, and most importantly communicate about them to the user properly, so the second option seems good.

**Comment by ashelkovnykov** [comment]
We already have options to control automatic updates per-app, and for the OS itself. It seems like the trick is finding the right balance. What about introducing the idea of "pinned" or "vital" apps, and then:

1. If `%base` updates disabled, never perform automatic updates
2. If `%base` updates enabled:
    1. If all "vital" apps support new kelvin, attempt upgrade once and revert to prev version if one upgarde fails
    2. Otherwise, prompt user to upgrade manually

If while waiting for `2.ii`, the conditions for `2.i` become met, then attempt the upgrade.

PS - We need the option to install old versions of apps, and to install them with automatic updates disabled (so that they don't update the second they're installed)

**Comment by ajlamarc** [comment]
Related request:
CLI-based app management tools:

|favor some-desk=@tas  (if a new %base OTA comes in with kelvin incompatibility to %some-desk, do not apply %base in order that %some-desk can keep running) (currently the default behavior, but really only important for mission-critical apps)

|demur some-desk=@tas  (if a new %base OTA comes in with kelvin incompatibility to %some-desk, suspend %some-desk so that the new %base can be applied) (this is so people don’t have important system updates silently put on the back burner just because some they downloaded an app from a friend four months ago that was never updated again)

|sched [some-desk=@tas hh=@ud mm=@ud ,day=[@ud]  (when a new update comes in for some-desk=@tas, don’t just apply the update, but wait until hh:mm to apply it (and optionally specify a list of days of the week).  So for example, |sched %lotus-1-2-3 [2 30 [6 0 ~]] would mean “if an update for %lotus-1-2-3 comes in, wait until 2:30am on Saturday or Sunday to apply it”.)

**Comment by philipcmonk** [comment]
I agree with the idea of introducing the concept of "vital" apps.  Further, I think the defaults are important, and that they should be:
- Any apps that you booted with are vital by default
- Apps you install afterward are not vital by default
- You can mark any app vital or non-vital in the app settings page (where you currently set the update source).  Obviously there should also be a generator.

Then, as long as the apps in the pill are kept up-to-date, most people will get kelvin updates immediately, unless they went out of their way to mark a particular app as vital.  I think this is better than the current state where every kelvin update they're asked the unusual question of whether they want to delay the update for whatever app they happen to have installed since the last kelvin.  Apps suspended in this way are automatically revived when the developer pushes a compatible update.

**Comment by Conviare** [comment]
FWIW previous to recent updates you could suspend and resume apps from landscape.  Now you cannot and appears as broken.  Can recover but sometime it appears you should not.
